### DISCUSSION TITLE
Running Zephyr on R7 core in RCAR H3(h3ulcb)

### DISCUSSION THREAD

**Discussion by u27p20** [zephyrproject-rtos/zephyr#67750] [Q&A]
Hello,

I am trying to flash and run Zephyr on the R7 core in the RCAR H3 Kingfisher Infotainment board. I've followed the steps mentioned in https://docs.zephyrproject.org/latest/boards/arm/rcar_h3ulcb/doc/rcar_h3ulcb.html.

I've successfully flashed the hello world application using "Olimex ARM-USB-OCD-H" and connected a USB to UART cable to CN04 on the Kingfisher board. The flashing logs indicate that the Zephyr binary is written to memory, but I've encountered some errors related to JTAG.
I have attached flashing logs using "west flash". PFA. 
[west-flash-logs.txt](https://github.com/zephyrproject-rtos/zephyr/files/13968880/west-flash-logs.txt)

After flashing the binary using "west flash", I don't see any serial prints on the serial console.

I'm currently stuck at this point and would like to determine whether the flashing using "west flash" was successful or not ? Additionally, I want to know if any specific configurations are required to run Zephyr on the R7 core in the RCAR H3 ?

Any assistance on this matter would be greatly appreciated.

Thanks in advance.

**Comment by ndrs-pst** [comment]
Perhaps @aaillet, who has had access to the hardware, could offer some insights.
Since the issue [#45168](https://github.com/zephyrproject-rtos/zephyr/issues/45168) on the Zephyr Project, involves the RCAR board.

Additionally, this link could be of interest to you: [R-Car H3ULCB CR7 Board - Hello World Demo on Renodepedia](https://renodepedia.renode.io/boards/rcar_h3ulcb_cr7/?view=software&demo=hello_world).

**Comment by aaillet** [comment]
Hi, sorry for late reply, first time in the Zephyr discussions tab x)
Thanks for pinging me @ndrs-pst.

I'll check if I can reproduce the issue ASAP.

Please @u27p20 can you give me some more information about your setup:
- H3ULCB version (see https://elinux.org/R-Car/Boards/H3SK#Hardware)
- Kingfisher version (S03-M03-M04-M05-M06)
- Software environment (Linux distribution ?)
- Zephyr version/branch on which you built the sample

Did you followed the complete Zephyr getting started guide ? (https://docs.zephyrproject.org/latest/develop/getting_started/index.html)

Depending on these information, some HW tweaks may be applied.

Thanks by advance for your answers

**Comment by u27p20** [comment]
Hi,

Thanks for your messages. 

I've found a solution. The issue was related to the bootloaders of the A5x core. I resolved it by updating to the latest bootloaders for A5x and successfully flashed the Zephyr RTOS. I can observe the serial log on the terminal for hello world example.

Currently, Zephyr is flashed at address 0x40040000 in RAM, which is volatile memory. I need to flash a binary after each reset. Is there a method to store binary files in non-volatile memory like hyper flash that persists even after a reset and boot from the non-volatile memory?

Thanks in advance.

**Reply by aaillet** [reply]
Hi,

Nice you found a solution, I would be interested to know which bootloader was creating this issue (and in which version).

For you to know, for interrupt driven driver to work fine, you need a patched ARM Trusted Firmware (see https://elinux.org/R-Car/Boards/Zephyr-Gen3/2.7#Build_.E2.80.9CARM_Trusted_Firmware.E2.80.9D_.28BL2.29).

Concerning flashing to non-volatile memory, from what I know, it's not currently possible to flash an RTOS to internal memory (using flashwriter) as bootloader would not run it at boot.

A good "automated" solution is to load and run Zephyr from SD Card at boot using U-Boot:

I'll use blinky case as example (https://docs.zephyrproject.org/latest/samples/basic/blinky/README.html)

- Copy your Zephyr binary to the `boot` partition of SD card (alongside Linux Image), if you want multiple binary to be available, rename Zephyr binary. (eg. `boot/blinky-demo.bin`)
  - Zephyr binary to use is located in `build_folder/zephyr/zephyr.bin`
- Add following environment variables to U-Boot
```
boot_cr7=mw.l 0xf1120004 0xff;mw.l 0xe6160070 ${fileaddr};setexpr.l cr7bar_val ${fileaddr} + 0x10;mw.l 0xe6160070 ${cr7bar_val};setexpr.l cr7bar_val ${fileaddr} + 0x10;mw.l 0xe6160070 ${cr7bar_val};mw.l 0xe6150904 0xa5a50000;mw.l 0xe618024c 0x1 && sleep 1 && mw.l 0xe61500b0 0x00400000;mw.l 0xe6150948 0x00400000
bootmmc=0:1
bootmode=mmc
bootzaddr=0x40040000
bootzfile=blinky-demo.bin
cr7=run load_cr7_${bootmode};run boot_cr7
cr7bar_val=40040010
fileaddr=40040000
load_cr7_mmc=ext4load mmc ${bootmmc} ${bootzaddr} ${bootzfile} && dcache flush
```
- Use `run cr7` command, in "blinky" case, middle LED (LED5) will start blinking.

To integrate it alongside Linux, you could prefix your `bootcmd` variable by `cr7;`

I just tested that solution on my H3ULCB(WS3.0)+KF(M06) setup and it worked , I hope it will answer your needs.

PS: another solution is to load Zephyr using remoteproc but it's not an easy solution to use.

**Reply by u27p20** [reply]
Hi Aaillet,

I wanted to update you on the  with the bootloaders first. The previous versions included changes to initiate the R7 core from A5x, but this configuration was causing issues. I found that flashing the bootloaders without these changes resolved the problem.

However, I've encountered a new challenge. While running the '`load_cr7_mmc`' command, the image doesn't load into the expected memory location '0x40040000,' and an error about reserved memory is generated. The error logs  are below.
```
=> run load_cr7_mmc
** Reading file would overwrite reserved memory **
Failed to load 'boot/zephyr.bin'
```

To troubleshoot this, I conducted a test : 

1.  Loading the binary file from the SD card into memory location '0x48080000' 
2. Copying binary file from  '0x48080000'  to '0x40040000.' 
3.  running '`load_cr7_mmc`' and '`boot_cr7`'.

 I observed that the binary file was successfully copied to '0x40040000' 
by using 'md' command. However, there was no output on the LED, and no serial logs appeared on the terminal.

I have attached the memory dump and outlined the steps taken in the attached document, PFA
[test_logs_and_steps.txt](https://github.com/zephyrproject-rtos/zephyr/files/14026672/test_logs_and_steps.txt)
. Despite copying the binary file to '0x40040000,' there seems to be no output after running '`boot_cr7`.'

I'm using the H3 starter kit (No.20203) and KF(M05). Any insights or guidance you can provide on resolving this issue would be greatly appreciated.

Thanks in advance

**Reply by aaillet** [reply]
Hi,

I'll try to find out what's going on ASAP.

Can you provide me the guide you followed to build and flash bootloaders ? (including yocto version used)

As said before, to be able to help you debug, I need you to use supported bootloader version, you at least need to rebuild and flash ARM Trusted Firmware with a custom patch, please follow this guide : https://elinux.org/R-Car/Boards/Zephyr-Gen3/2.7#Build_.E2.80.9CARM_Trusted_Firmware.E2.80.9D_.28BL2.29

Please give me as much information as possible so I will be able to help you

**Reply by u27p20** [reply]
Hi,

First, let me explain the steps that I have followed : 

1. I have used the build script from  Kingfisher/Yocto-Gen3/v5.9.0 (https://elinux.org/R-Car/Boards/Kingfisher/Yocto-Gen3/v5.9.0).
2. included  meta-rcar-zephyr 
```
git clone https://github.com/iotbzh/meta-rcar-zephyr
checkout -b tmp origin/cr7_5.5.0_and_later
bitbake-layers add-layer ../../meta/meta-rcar-zephyr/
```
3.  Build ATF images : "bitbake arm-trusted-firmware"
4.  Flashing the bootloader I have referred: https://elinux.org/R-Car/Boards/H3SK#Flashing_firmware
5.  I  have built the latest zephyr(3.5.99) binary by referring the  zephyr documentation (https://docs.zephyrproject.org/latest/develop/getting_started/index.html)
6. I referred your previous comment to boot zephyr from SD card. 

I noticed two differences from the elinux page I followed  and with the Zephyr-Gen3/2.7 (https://elinux.org/R-Car/Boards/Zephyr-Gen3/2.7#Build_.E2.80.9CARM_Trusted_Firmware.E2.80.9D_.28BL2.29). 

1. The elinux page that I have followed  has a different meta-renesas commit ID, and the meta-rcar layer is included.
2. I used the latest version of Zephyr.

These are the detailed steps that I followed to build the ATF and Zephyr. Please let me know if any further details are required.

In the meantime, I will build and flash the ATF and Zephyr by following the steps outlined at https://elinux.org/R-Car/Boards/Zephyr-Gen3/2.7#Build_.E2.80.9CARM_Trusted_Firmware.E2.80.9D_.28BL2.29.


Thank you.

**Reply by u27p20** [reply]
Hi,

I've successfully built the bootloader images and Zephyr 2.7 by following the instructions on the eLinux pages (https://elinux.org/R-Car/Boards/Zephyr-Gen3/2.7#Build_.E2.80.9CARM_Trusted_Firmware.E2.80.9D_.28BL2.29). After flashing the bootloaders, I attempted to boot Zephyr from the SD card, but encountered the same error as before.

```
=> run  load_cr7_mmc
** Reading file would overwrite reserved memory **
Failed to load 'boot/zephyr.bin'

```
I'm not  understanding how it works at your end and why it's not working with my setup.

**Reply by aaillet** [reply]
Hi,
I'm working on fixing your issue.
At first glance, we don't use the same U-Boot version, it looks like issue is coming from that.
I will get back to you as soon as I have an answer.
For now, please flash Zephyr using `west flash`.
Sorry for that

**Reply by u27p20** [reply]
Hi @aaillet 

May I know is there any  update on booting Zephyr from the SD card?

I also have another question regarding peripheral sharing between Linux and Zephyr. 

I have a custom board where I've flashed Zephyr using Olimex-ARM-USB-OCD-H, and U-boot & Linux use SICF2 for serial log. Since my custom board lacks an extra serial interface for Zephyr, I configured SCIF2 for serial log in the Zephyr device tree. Both Linux and Zephyr use SCIF2 in my setup.

Initially, I assumed that only Linux or Zephyr couldn't use the same peripheral at the same time, and I expected to encounter failures during booting.

 However, after flashing the Zephyr image, I observed both Linux and Zephyr uses SCIF2, and the logs of Helloworld were printed along with  U-boot logs.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/112854936/42f97318-b5c8-4dfe-967c-bbe4324c77bb)


I'm puzzled about how Linux and Zephyr share the same peripherals. I couldn't find any information in the Zephyr documentation about how peripherals are shared between two operating systems. Could you please provide insights into how SCIF2 is shared between Linux and Zephyr?


Thanks in advance

**Reply by aaillet** [reply]
Hi @u27p20 ,

No updates yet on how to boot Zephyr from SD card (it's confirmed that it may works depending on uboot version used).

Concerning serial lines sharing, you may have seen that documented/default setup is the following:
- SCIF2 is used by U-Boot then Linux (or any OS running on A57 cores)
- SCIF1 is used by Zephyr on Kingfisher CN04  (or H3ULCB if CN10 is welded)

Theoretically, you can use any peripheral in both worlds at the same time, there is no hardware barrier for both CA57 and CR5 cores to read/write registers of any controller "at the same time".
However it won't work correctly as there is obviously only one pair of interrupts per serial line. may be catch by the wrong core at the wrong time, etc (eg .

Above explanation is also true for other peripherals, as any peripheral accessible by A57 cores can be accessed by CR7 core.
The recommended way to work with both worlds is to dispatch peripherals to one or other device tree.

In serial line case, different configuration (baudrate...) may also introduce incorrect characters when reading serial port.
Even with the same configuration, when both worlds are writing to the same time, interrupts issues may happen.

For your custom case, I don't really know what to recommend, here a few possibilities to explore:
- Confirm zephyr behavior using scif then keep serial line for ca57 cores as a final solution
- Configure your zephyr binary to print in ram_console and read it using gdb through olimex probe
- Explore the "trace" capabilities of remoteproc and read trace from linux.
  - Unfortunately,  there is not official support for remoteproc on h3ulcb zephyr port (but old dev branch could be used as a base)
  - A basic remoteproc driver has been upstreamed some time ago on linux side (see https://marc.info/?a=160760621600001&r=1&w=2)

I hope it answer some of your explanation.
Regards,
Aymeric
